### PR TITLE
Allow exposing some initial provider config options via web site

### LINKED ISSUE #10571: Allow exposing some initial provider config options via web site
### Description

In order to better document provider configuration options we should be adding a page in our web site to list some key provider configuration options.

### Discussion

_No response_

### Motivation

Make easier to format provider configuration options when migrating from the legacy distribution.

### Details

_No response_

### PR DESCRIPTION
**Pull request by pedroigor**
Closes #10571

* Adds a method to obtain configuration metadata from `ProviderFactory`
* Updates some key providers to expose their configuration metadata
* Adds a guide in our web site to list all the available provider configuration options

### CODE REVIEW

**Review by DGuhr — DISMISSED** [review]
lgtm in general, thank you very much.👍 As i dont have an overview/opinion over all spi config that should be exposed here but maybe isnt included, i'd suggest @stianst also reviews it with a focus on that, so we don't forget things.

**Comment by pedroigor** [comment]
@DGuhr I did not have about all of them too so I checked the hierarchy for all implementations that exposes configuration. I tried to not include implementations that are already wrapped in our config or configuration that I was not sure if we want to make public.

File: model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/lock/LiquibaseDBLockProviderFactory.java [hunk 1]
```diff
@@ -74,4 +77,14 @@ public String getId() {
...
         return PROVIDER_PRIORITY;
     }
+
+    @Override
+    public List<ProviderConfigProperty> getConfigMetadata() {
+        return ProviderConfigurationBuilder.create()
+                .property()
+                .name("openid-configuration-override")
+                .type("string")
+                .helpText("The file path from where the metadata should be loaded from. You can use an absolute file path or, if the file is in the server classpath, use the 'classpath:' prefix to load the file from the classpath.")
```
Comment (stianst) [inline model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/lock/LiquibaseDBLockProviderFactory.java:87]:
```suggestion
                .name("lockWaitTimeout")
                .type("int")
                .helpText("The maximum time to wait when waiting to release a database lock.")
```

File: services/src/main/java/org/keycloak/connections/httpclient/DefaultHttpClientFactory.java [hunk 2]
```diff
@@ -225,6 +228,86 @@ public void postInit(KeycloakSessionFactory factory) {
...
+                .name("client-key-password")
+                .type("string")
+                .helpText("The key password.")
+                .defaultValue(-1L)
+                .add()
+                .property()
+                .name("disable-trust-manager")
+                .type("boolean")
+                .helpText("Disable trust management and hostname verification. NOTE this is a security hole, so only set this option if you cannot or do not want to verify the identity of the host you are communicating with.")
+                .defaultValue(false)
```
Comment (stianst) [inline services/src/main/java/org/keycloak/connections/httpclient/DefaultHttpClientFactory.java:301]: If not mistaken truststore for the http-client should be configured in the truststore provider, and not directly in the http-client provider? So, perhaps we should not include these?

File: services/src/main/java/org/keycloak/executors/DefaultExecutorsProviderFactory.java [hunk 3]
```diff
@@ -100,6 +103,24 @@ public String getId() {
...
+                .defaultValue(DEFAULT_MIN_THREADS)
+                .add()
+                .property()
+                .name("<task-type>-max")
+                .type("int")
+                .helpText("The maximum number of threads to allow in the pool.")
+                .defaultValue(DEFAULT_MAX_THREADS)
+                .add()
+                .build();
+    }
```
Comment (stianst) [inline services/src/main/java/org/keycloak/executors/DefaultExecutorsProviderFactory.java:122]: Wonder if this is useful at all as the `task types` are not defined anywhere. This whole thing is a bit messy to be honest, and also quite crazy that we have potentially up to 64 x <several types> threads running in the background doing background jobs. Honestly got no clue why we even have multiple different pools for background tasks.

Perhaps just leave this one out for now?

File: services/src/main/java/org/keycloak/protocol/oidc/OIDCWellKnownProviderFactory.java [hunk 4]
```diff
@@ -88,6 +91,24 @@ public int getPriority() {
...
     }
 
+    @Override
+    public List<ProviderConfigProperty> getConfigMetadata() {
+        return ProviderConfigurationBuilder.create()
+                .property()
+                .name("lockWaitTimeout")
+                .type("long")
+                .helpText("The maximum time to wait when waiting to release a database lock.")
+                .defaultValue(900L)
```
Comment (stianst) [inline services/src/main/java/org/keycloak/protocol/oidc/OIDCWellKnownProviderFactory.java:101]:
```suggestion
                .name("openid-configuration-override")
                .type("string")
                .helpText("The file path from where the metadata should be loaded from. You can use an absolute file path or, if the file is in the server classpath, use the 'classpath:' prefix to load the file from the classpath.")
```

**Comment by abstractj** [comment]
@pedroigor could you please rebase?

**Comment by pedroigor** [comment]
@abstractj Thanks, rebased. Let us now wait for @stianst feedback to make sure the changes to the website are good enough.

**Review by stianst — CHANGES_REQUESTED** [review]
There's quite a bit of missing stuff, not sure really what should be included or not. 

Was looking for config usage by looking at what uses Config.Scope methods and found stuff like:

* BlacklistPasswordPolicyProviderFactory
* Theme config (DefaultThemeManager), but that one is a bit awkward as it's used by the manager rather than providers
* FileTruststoreProviderFactory
* JSPolicyProviderFactory
* JBossLoggingEventListenerProviderFactory
* JpaClientProviderFactory
* SAMLIdentityProviderFactory
* EmailEventListenerProviderFactory
* SamlProtocolFactory
* AbstractUserProfileProvider
* AbstractClientCertificateFromHttpHeadersLookupFactory
* BasicTimerProviderFactory
* Loads of stuff around Infinispan

From a priority of the above perhaps we should include:

* Event listeners
* Theme config
* Truststore

By the way did you take a look at the documentation (old server install and admin guides) to see what properties we mention there? At the very least we should probably include everything that is mentioned through documentation, as that would suggest they are somewhat priorities ;)

**Comment by pedroigor** [comment]
Thanks, let me start by working on the suggestions that I can initially include in this PR.

File: services/src/main/java/org/keycloak/connections/httpclient/DefaultHttpClientFactory.java (hunk 2 above)
Comment (pedroigor) [reply services/src/main/java/org/keycloak/connections/httpclient/DefaultHttpClientFactory.java:301]: Yes, it is.

But the `client-keystore-*` options are related to the key material (private key mainly) to set up TLS for outbound connections.

And the `disable-trust-manager` is still used to explicitly disable the trustmanager when building the client config. We can evaluate later if this option is redundant but for now, I would include it.

**Comment by pedroigor** [comment]
@stianst Does it make sense to include a brief description for each SPI and Provider?

**Comment by stianst** [comment]
Couple things:

* Type is defined in the providers, but not shown in the docs
* "Build-icon" column is shown even though it's not relevant to provider options

Sent a PR to fix the above here:
https://github.com/pedroigor/keycloak/pull/6

**Review by stianst — COMMENTED** [review]
LGTM, but would probably be nice to include the PR I sent here https://github.com/pedroigor/keycloak/pull/6

**Comment by pedroigor** [comment]
@stianst Commit added. Thanks for the changes.

**Review by stianst — APPROVED** [review]

**Comment by emmansun** [comment]
It appears that we are unable to locate this item on the Keycloak blog archive at https://www.keycloak.org/blog-archive . Could you please assist in verifying whether this enhancement has been supported since version 18.0.0?

**Pull request merged**